src/lava/lib/optimization/solvers/generic/builder.py

```diff
@@ -140,55 +142,73 @@ def _create_process_constructor(
             self,
+            backend,
             hyperparameters: ty.Dict[str, ty.Union[int, npt.ArrayLike]],
             name: ty.Optional[str] = None,
             log_config: ty.Optional[LogConfig] = None,
         ) -> None:
             super(type(self), self).__init__(
+                backend=backend,
                 hyperparameters=hyperparameters,
                 name=name,
                 log_config=log_config,
             )
             self.problem = problem
             self.hyperparameters = hyperparameters
+            self.backend = backend
+            self.is_continuous = 0
+            self.is_discrete = 0
             if not hasattr(problem, "variables"):
                 raise Exception(
                     "An optimization problem must contain " "variables."
                 )
-            if hasattr(problem.variables, "continuous") or isinstance(
-                problem.variables, ContinuousVariables
+            if (
+                hasattr(problem.variables, "continuous")
+                and problem.variables.continuous.num_variables is not None
             ):
                 self.continuous_variables = Var(
-                    shape=(problem.variables.continuous.num_vars, 2)
+                    shape=(problem.variables.continuous.num_variables,)
                 )
-            if hasattr(problem.variables, "discrete") or isinstance(
-                problem.variables, DiscreteVariables
+                self.is_continuous = 1
+            if (
+                hasattr(problem.variables, "discrete")
+                and problem.variables.discrete.num_variables is not None
             ):
                 self.discrete_variables = Var(
                     shape=(
-                        problem.variables.num_variables,
+                        problem.variables.discrete.num_variables,
                         # problem.variables.domain_sizes[0]
                     )
                 )
+                self.is_discrete = 1
             self.cost_diagonal = None
             if hasattr(problem, "cost"):
                 mrcv = SolverProcessBuilder._map_rank_to_coefficients_vars
                 self.cost_coefficients = mrcv(problem.cost.coefficients)
-                self.cost_diagonal = problem.cost.coefficients[2].diagonal()
-            self.variable_assignment = Var(
-                shape=(problem.variables.num_variables,)
-            )
-            self.best_variable_assignment = Var(
-                shape=(problem.variables.num_variables,)
-            )
-            # Total cost = optimality_first_byte << 24 + optimality_last_bytes
-            self.optimality_last_bytes = Var(shape=(1,))
-            self.optimality_first_byte = Var(shape=(1,))
-            self.optimum = Var(shape=(2,))
-            self.feasibility = Var(shape=(1,))
-            self.solution_step = Var(shape=(1,))
-            self.cost_monitor = Var(shape=(1,))
+                if self.is_discrete:

```

Comment:
Does is_discrete=1 imply that the cost is quadratic? It does currently (QUBO), but that is unlikely to be general.

Reply:
![image](https://github.com/lava-nc/lava-optimization/assets/12975862/a59602d7-739f-408c-81e2-df1bb252b4b6)
is_discrete is set when the problem has discrete variables. 
Keep in mind that the code has to be refactored as more solvers are added in the future. Making it too generic is a huge effort in terms of planning and software engineering for which I don't have the bandwidth atm.

Reply:
will remove the ghost code in an upcoming commit

Reply:
Just noticed that you also use the self.is_continuous variable.
Maybe worth renaming them to
self.has_continuous_variables
self.has_continuous_variables?

Reply:
made change